otter/convergence/service.py

```diff
@@ -1,17 +1,95 @@
 """Converger service"""
 
+import time
+
 from functools import partial
 
 from effect import Effect
 from effect.twisted import perform
 
+from toolz.itertoolz import concat
+
 from twisted.application.service import Service
 
 from otter.constants import CONVERGENCE_LOCK_PATH
-from otter.convergence.composition import (
-    execute_convergence, get_desired_group_state)
+from otter.convergence.composition import get_desired_group_state
+from otter.convergence.effecting import steps_to_effect
+from otter.convergence.gathering import get_all_convergence_data
+from otter.convergence.model import ServerState
+from otter.convergence.planning import plan
 from otter.http import TenantScope
+from otter.models.intents import ModifyGroupState
 from otter.util.deferredutils import with_lock
+from otter.util.fp import assoc_obj
+
+
+def server_to_json(server):
+    """
+    Convert a NovaServer to a dict representation suitable for returning to the
+    end-user as a part of group state.
+    """
+    return {'id': server.id}
+
+
+def determine_active(servers, lb_nodes):
+    """
+    Given the current NovaServers and LB nodes, determine which servers are
+    completely built.
+
+    :param servers: sequence of :obj:`NovaServer`.
+    :param lb_nodes: sequence of :obj:`ILBNode`.
+
+    :return: list of servers that are active.
+    """
+
+    def all_met(server, current_lb_nodes):
+        """Determine if a server is in all the LBs it wants to be in."""
+        desired_lbs = set(concat(server.desired_lbs.values()))
+        met_desireds = set([
+            desired for desired in desired_lbs
+            for node in current_lb_nodes
+            if desired.equivalent_definition(node.description)])
+        return desired_lbs == met_desireds
+
+    return [s for s in servers
+            if s.state == ServerState.ACTIVE
+            and all_met(s, [node for node in lb_nodes if node.matches(s)])]
+
+
+def execute_convergence(
+        scaling_group, desired, launch_config, now, log,
+        get_all_convergence_data=get_all_convergence_data):
+    """
+    Gather data, plan a convergence, save active and pending servers to the
+    group state, and then execute the convergence.
+
+    :param IScalingGroup scaling_group: The scaling group object.
+    :param GroupState group_state: The group state.
+    :param launch_config: An otter launch config.
+    :param now: The current time in seconds.
+    :param log: A bound logger.
+    :param get_all_convergence_data: The :func`get_all_convergence_data` to use
+        for testing.
+
+    :return: An Effect of List of Step Effect Results. (???)
```

Comment:
Oh right, I forgot about that. The problem is that we still don't have the (StepResult, [String]) thing happening. I'll tweak this docstring though.